Valentine Kopotchel/0/bread_bot.rb

```diff
@@ -0,0 +1,321 @@
+  '2018-01' => 'data/Average_prices(serv)-01-2018.xlsx',
+  '2018-02' => 'data/Average_prices(serv)-02-2018.xlsx',
+  '2018-03' => 'data/Average_prices(serv)-03-2018.xlsx',
+  '2018-04' => 'data/Average_prices(serv)-04-2018.xlsx',
+  '2018-05' => 'data/Average prices(serv)-05-2018.xlsx',
+  '2018-06' => 'data/Average_prices(serv)-06-2018.xlsx',
+  '2018-07' => 'data/Average prices(serv)-07-2018.xlsx',
+  '2018-08' => 'data/Average_prices(serv)-08-2018.xlsx',
+  '2018-09' => 'data/Average_prices(serv)-09-2018.xlsx',
+  '2018-10' => 'data/Average_prices(serv)-10-2018.xlsx'
+}.freeze
+# Doing some dirty work to make less code in BreadBot
+class BotHelper
+  PRICE_COLUMN_MINSK = 16
+  NAME_COLUMN = 0
+
+  def build_page(key)
+    path = HASH_OF_TABLES[key]
+    xlsx = Roo::Spreadsheet.open(path)
+    xlsx.sheet 0
+  end
+
+  def deflation?(date)
+    date[0, 4] < '2017'
+  end
+
+  def exists?(arg)
+    arg.to_s.length.positive?
+  end
+
+  def num?(arg)
+    arg.is_a? Numeric
+  end
+
+  def build_hash(key)
+    fin_hash = {}
+    condition = deflation? key
+    build_page(key).each do |row|
+      next unless exists?(row[NAME_COLUMN]) && num?(row[PRICE_COLUMN_MINSK])
+
+      val = row[PRICE_COLUMN_MINSK]
+      val /= 10_000.0 if condition
+      fin_hash[row[NAME_COLUMN]] = val
+    end
+    fin_hash
+  end
+
+  def make_normal(str)
+    str.swapcase.capitalize
+  end
+
+  def reg_build(str)
+    str = str.strip
+    r = str.upcase
+    temp = Regexp.new("(#{r}$|#{r}\s|#{r},)")
+    Regexp.and_union(/(^|[\P{L}])/, temp)
+  end
+
+  def split_key(str)
+    [str[0, 4], str[-2, 2]]
+  end
+
+  def minmax_assign(hash)
+    min = hash.keys.min
+    max = hash.keys.max
+    a = hash[min]
+    b = hash[max]
+    [min, max, a, b]
+  end
+
+  def write_minmax_info(hash)
+    min, max, a, b = minmax_assign(hash)
+    puts "Lowest was on #{(split_key a)[1]}/#{(split_key a)[0]} "\
+"at #{min.round(2)} BYN"
+    puts "Highest was on #{(split_key b)[1]}/#{(split_key b)[0]} "\
+"at #{max.round(2)} BYN"
+  end
+end
+# The main class
+class BreadBot
+  def initialize
+    @big_hash = {}
+    dog = BotHelper.new
+    HASH_OF_TABLES.each_key do |key|
+      @big_hash[key] = dog.build_hash key
+    end
+    @h = @big_hash[@big_hash.keys.max]
+  end
+
+  def gen_string(key)
+    ' is ' + @h[key].round(2).to_s + 'BYN in Minsk nowadays'
+  end
+
+  def count(key, str)
+    h = {}
+    @big_hash.each_key do |date|
+      val = get_price_by_date(date, key, str)
+      h[date] = val if val > -1
+    end
+    min = h.values.min
+    max = h.values.max
+    { min => h.key(min), max => h.key(max) }
+  end
+
+  def get_price_by_date(date, key, str)
+    work_hash = hash[date]
+    return work_hash[key] if work_hash.key? key
+
+    a = find_match(BotHelper.new.reg_build(str), work_hash)
+    return work_hash[a[0]] if a.size == 1
+
+    -1
+  end
+
+  def hash
+    @big_hash
+  end
+
+  def find_match(reg, hash = @h)
+    a = []
+    hash.each_key do |key|
```

Comment:
you can just return hash.each_key.select { ... } without creating temporary array